cli/src/commands/git/colocation.rs

```diff
@@ -0,0 +1,343 @@
+// You may obtain a copy of the License at
+//
+// https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::io::Write as _;
+
+use jj_lib::file_util::IoResultExt as _;
+use jj_lib::git;
+use jj_lib::git::get_git_repo;
+use jj_lib::object_id::ObjectId as _;
+use jj_lib::op_store::RefTarget;
+use jj_lib::repo::Repo as _;
+
+use crate::cli_util::CommandHelper;
+use crate::command_error::CommandError;
+use crate::command_error::user_error;
+use crate::command_error::user_error_with_message;
+use crate::commands::git::maybe_add_gitignore;
+use crate::git_util::is_colocated_git_workspace;
+use crate::ui::Ui;
+
+#[derive(clap::Args, Clone, Debug)]
+pub struct GitColocationEnableArgs {}
+
+#[derive(clap::Args, Clone, Debug)]
+pub struct GitColocationDisableArgs {}
+
+#[derive(clap::Args, Clone, Debug)]
+pub struct GitColocationStatusArgs {}
+
+/// Manage Jujutsu repository colocation with Git
+#[derive(clap::Subcommand, Clone, Debug)]
+pub enum GitColocationCommand {
+    /// Enable colocation (convert into a colocated Jujutsu/Git repository)
+    ///
+    /// This moves the underlying Git repository that is found inside the .jj
+    /// directory to the root of the Jujutsu workspace. This allows you to
+    /// use Git commands directly in the Jujutsu workspace.
+    Enable(GitColocationEnableArgs),
+    /// Disable colocation (convert into a non-colocated Jujutsu/Git
+    /// repository)
+    ///
+    /// This moves the Git repository that is at the root of the Jujutsu
+    /// workspace into the .jj directory. Once this is done you will no longer
+    /// be able to use Git commands directly in the Jujutsu workspace.
+    Disable(GitColocationDisableArgs),
+    /// Show the current colocation status
+    Status(GitColocationStatusArgs),
+}
+
+pub fn cmd_git_colocation(
+    ui: &mut Ui,
+    command: &CommandHelper,
+    subcommand: &GitColocationCommand,
+) -> Result<(), CommandError> {
+    let workspace_command = command.workspace_helper(ui)?;
+
+    // Check if backend is Git
+    if git::get_git_backend(workspace_command.repo().store()).is_err() {
+        return Err(user_error(
+            "This command requires a repository backed by Git. This repository appears to be \
+             using a different backend.",
+        ));
+    }
+
+    // Ensure that this is the main workspace
+    let workspace_root = workspace_command.workspace_root();
+    let jj_repo_path = workspace_root.join(".jj").join("repo");
+    if jj_repo_path.is_file() {
+        return Err(user_error(
+            "This command cannot be used in a non-main Jujutsu workspace.",
+        ));
+    }
```

Comment:
Perhaps, this can be extracted to helper function and let `cmd_git_colocation_*()` call it. It's weird that we load the workspace twice.